### PR TITLE
[mesh] mesh frame editing part-1 (qgis_core)

### PR DESCRIPTION
**Pull request by vcloarec**
This PR is the first step of the mesh frame editing implementation.
Here, some core classes are introduced to:
* ensure consistency of the mesh frame during editing
* retrieve easily mesh faces from mesh vertices (before this PR, it was only possible to retrieve vertices from faces without using spatial index)
* add/remove face/vertices on the mesh with undo/redo

The principal new classes are :
* QgsMeshEditor, exposed in Python, that receive the edit request and is responsible for the synchronization with the triangular mesh that is used for displaying the mesh
* QgsTopologicalMesh : A wrapper of the native mesh creates topological information on the mesh when editing starts. This information is essentially neighboring of faces and a map from vertices to faces. This class has the responsibility to keep the mesh topologically correct and to operate modification on the mesh. This class is not exposed to Python.

For now, the modification of the mesh is only available from the Python API, and saving it is not already implemented. The implementation is largely tested but surely not completely stable for now.

mesh_editing_1

### CODE REVIEW

**Comment by PeterPetrik** [comment]
first part of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/228

File: src/core/mesh/qgsmeshdataset.cpp [hunk 1]
```diff
@@ -1071,3 +1072,98 @@ void QgsMeshDatasetGroup::setIsScalar( bool isScalar )
...
+{
+  if ( index != 0 )
+    return nullptr;
+
+  return mDataset.get();
+}
+
+QgsMeshDatasetGroup::Type QgsMeshVerticesElevationDatasetGroup::type() const
+{
+  return QgsMeshDatasetGroup::Memory; //maybe create a new type ?
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataset.cpp:1167]: I think memory is ok...

File: src/core/mesh/qgstriangularmesh.h [hunk 2]
```diff
@@ -206,6 +228,70 @@ class CORE_EXPORT QgsTriangularMesh // TODO rename to QgsRendererMesh in QGIS 4
...
+        QList<int> mNativeFaceIndexesToRemove;
+        QVector<QgsMeshFace> mNativeFacesToRemove;
+
+        mutable int mTrianglesAddedStartIndex = 0;  // defined each time the changes are apply
+        mutable QList<int> mRemovedTriangleIndexes; // defined when changes are apply the first time
+
+        friend class QgsTriangularMesh;
+    };
+
+    /*!
```
Comment (PeterPetrik) [inline src/core/mesh/qgstriangularmesh.h:265]: we use /** notation, can you please fix in the whole PR?

File: src/core/mesh/qgstriangularmesh.cpp [hunk 3]
```diff
@@ -611,3 +676,118 @@ QSet<int> QgsMeshUtils::nativeVerticesFromEdges( const QList<int> &edgesIndexes,
...
+
+  // remove vertices
+  // for now, let's try to not remove the vertices, because if the vertex is not referenced in faces,
+  // there is no access anymore to the vertex. If we do not remove it, not need to store (x,y,z) in the changes instance
+}
+
+void QgsTriangularMesh::reverseChanges( const QgsTriangularMesh::Changes &changes )
+{
+  //reverse added faces and added vertices
+  if ( !changes.mNativefacesToAdd.isEmpty() )
```
Comment (PeterPetrik) [inline src/core/mesh/qgstriangularmesh.cpp:750]: mNativefacesToAdd -> mNativeFacesToAdd

File: src/core/mesh/qgstopologicalmesh.h [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+};
+
+/**
+ * \ingroup core
+ *
+ * \brief  Convenient class that turn around a vertex and provide information about faces and vertices
+ *
+ * \since QGIS 3.22
+ */
+class QgsMeshVertexCirculator
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.h:226]: move to qgsmeshvertexcirculator.cpp[h].

Also it is not clear from the class name and description what the class does. is it described in qep somewhere? maybe topologicalMeshTraveller? not sure

File: src/core/mesh/qgstopologicalmesh.h [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+     */
+    Changes removeVertex( int vertexIndex, bool fillHole = false );
+
+    //! Applies the changes
+    void applyChanges( const Changes &changes );
+
+    //! Reverses the changes
+    void reverseChanges( const Changes &changes );
+
+    //! Check the topology of the face and set it counter clock wise if necessary
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.h:190]: Checks ... and sets

File: src/core/mesh/qgstopologicalmesh.h [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    Changes removeFaces( const QList<int> facesIndexes );
+
+    /**
+     *  Add a \a vertex in the face with index \a faceIndex. The including face is removed and new faces surrounding the added vertex are added.
+     *  The method returns a instance of the class QgsTopologicalMesh::Change that can be used to reverse or reapply the operation.
+     */
+    Changes addVertexInface( int faceIndex, const QgsMeshVertex &vertex );
+
+    /**
+     *  Add a free \a vertex in the face, that is a vertex tha tis not included or linked with any faces.
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.h:171]: Adds

File: src/core/mesh/qgstopologicalmesh.h [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    QgsMeshEditingError canFacesBeRemoved( const QList<int> facesIndexes );
+
+    /**
+     *  Removes faces with index in \a faceIndexes.
+     *  The method returns a instance of the class QgsTopologicalMesh::Change that can be used to reverse or reapply the operation.
+     */
+    Changes removeFaces( const QList<int> facesIndexes );
+
+    /**
+     *  Add a \a vertex in the face with index \a faceIndex. The including face is removed and new faces surrounding the added vertex are added.
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.h:165]: Adds

File: src/core/mesh/qgstopologicalmesh.h [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+SIP_NO_FILE
+
+class QgsMeshEditingError;
+class QgsMeshVertexCirculator;
+
+/**
+ * \ingroup core
+ *
+ * \brief Class that wraps a QgsMesh to ensure the consistency of the mesh during editing and help to access to elements from other elements
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.h:29]: what is "element" face/edge?

File: src/core/mesh/qgsmeshspatialindex.cpp [hunk 9]
```diff
@@ -382,3 +382,41 @@ QgsMesh::ElementType QgsMeshSpatialIndex::elementType() const
...
+
+void QgsMeshSpatialIndex::addFace( int faceIndex, const QgsMesh &mesh )
+{
+  if ( mesh.face( faceIndex ).isEmpty() )
+    return;
+  SpatialIndex::Region r( faceToRegion( mesh, faceIndex ) );
+
+  QMutexLocker locker( &d->mMutex );
+
+  // TODO: handle possible exceptions correctly
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshspatialindex.cpp:394]: todo

File: src/core/mesh/qgsmeshlayer.cpp [hunk 10]
```diff
@@ -863,6 +870,75 @@ qint64 QgsMeshLayer::datasetRelativeTimeInMilliseconds( const QgsMeshDatasetInde
   return mDatasetGroupStore->datasetRelativeTime( index );
 }
 
+bool QgsMeshLayer::startFrameEditing( const QgsCoordinateTransform &transform )
+{
+  if ( mMeshEditor )
+  {
+    //TODO : sends a message to the log
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:877]: todo

File: src/core/mesh/qgsmeshlayer.cpp [hunk 11]
```diff
@@ -863,6 +870,75 @@ qint64 QgsMeshLayer::datasetRelativeTimeInMilliseconds( const QgsMeshDatasetInde
...
+
+  mMeshEditor = new QgsMeshEditor( this );
+
+  QgsMeshEditingError error = mMeshEditor->initialize();
+
+  if ( error.errorType != QgsMeshEditingError::NoError )
+  {
+    mMeshEditor->deleteLater();
+    mMeshEditor = nullptr;
+    //TODO : sends the error to the log
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:891]: todo

File: src/core/mesh/qgsmeshdataset.h [hunk 12]
```diff
@@ -759,6 +760,54 @@ class CORE_EXPORT QgsMeshMemoryDatasetGroup: public QgsMeshDatasetGroup
...
 };
 
+/**
+ * \ingroup core
+ *
+ * \brief Class that represents a dataset with elevation value of the vertices of a existing mesh that can be edited
+ *
+ * \since QGIS 3.22
+ */
+class QgsMeshVerticesElevationDataset: public QgsMeshDataset
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataset.h:770]: add QgsMeshVerticesElevationDataset[Group] classes  to new file

File: src/core/mesh/qgstopologicalmesh.h (hunk 4 above)
Comment (vcloarec) [reply src/core/mesh/qgstopologicalmesh.h:226]: not sure for the move, it is a quite particular class that is completely link to the topological mesh.
class inspired by: https://doc.cgal.org/latest/Circulator/index.html#circulatorsCirculators

File: src/core/mesh/qgstopologicalmesh.h (hunk 8 above)
Comment (vcloarec) [reply src/core/mesh/qgstopologicalmesh.h:29]: here elements are vertex/faces (edge not yet supported), the class allow  vertices <--> faces access in both direction.

File: src/core/mesh/qgsmeshdataset.cpp [hunk 13]
```diff
@@ -1071,3 +1072,98 @@ void QgsMeshDatasetGroup::setIsScalar( bool isScalar )
 {
   mIsScalar = isScalar;
 }
+
+
+QgsMeshVerticesElevationDataset::QgsMeshVerticesElevationDataset( QgsMesh *mesh ):
+  mMesh( mesh )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmeshdataset.cpp:1078]:
```suggestion
QgsMeshVerticesElevationDataset::QgsMeshVerticesElevationDataset( QgsMesh *mesh )
  : mMesh( mesh )
```

File: src/core/mesh/qgsmeshdataset.h (hunk 12 above)
Comment (nyalldawson) [inline src/core/mesh/qgsmeshdataset.h:770]: missing CORE_EXPORT

File: src/core/mesh/qgsmeshdataset.h [hunk 14]
```diff
@@ -759,6 +760,54 @@ class CORE_EXPORT QgsMeshMemoryDatasetGroup: public QgsMeshDatasetGroup
...
+{
+  public:
+    //! Consructor
```
Comment (nyalldawson) [inline src/core/mesh/qgsmeshdataset.h:773]: typo

File: src/core/mesh/qgsmeshdataset.h [hunk 15]
```diff
@@ -759,6 +760,54 @@ class CORE_EXPORT QgsMeshMemoryDatasetGroup: public QgsMeshDatasetGroup
...
+
+/**
+ * \ingroup core
+ *
+ * \brief Class that represents a dataset group with elevation value of the vertices of a existing mesh that can be edited
+ *        This dataset group contains only one dataset.
+ *
+ * \since QGIS 3.22
+ */
+class QgsMeshVerticesElevationDatasetGroup : public QgsMeshDatasetGroup
```
Comment (nyalldawson) [inline src/core/mesh/qgsmeshdataset.h:794]: missing CORE_EXPORT

File: src/core/mesh/qgsmesheditor.cpp [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+#include "qgsmeshlayer.h"
+#include "qgslogger.h"
+
+#include <poly2tri.h>
+
+#include <QSet>
+
+
+QgsMeshEditor::QgsMeshEditor( QgsMeshLayer *meshLayer ):
+  QObject( meshLayer )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:29]:
```suggestion
QgsMeshEditor::QgsMeshEditor( QgsMeshLayer *meshLayer )
  : QObject( meshLayer )
```

File: src/core/mesh/qgsmesheditor.cpp [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+  , mMesh( meshLayer ? meshLayer->nativeMesh() : nullptr )
+  , mTriangularMesh( meshLayer ? meshLayer->triangularMeshByLodIndex( 0 ) : nullptr )
+  , mUndoStack( meshLayer ? meshLayer->undoStack() : nullptr )
+{}
+
+QgsMeshEditor::QgsMeshEditor( QgsMesh *nativeMesh, QgsTriangularMesh *triangularMesh, QObject *parent ):
+  QObject( parent )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:36]:
```suggestion
QgsMeshEditor::QgsMeshEditor( QgsMesh *nativeMesh, QgsTriangularMesh *triangularMesh, QObject *parent )
  : QObject( parent )
```

File: src/core/mesh/qgsmesheditor.cpp [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+
+void QgsMeshEditor::stopEditing()
+{
+  mTopologicalMesh.reindex();
+}
+
+
+
+QgsMeshLayerUndoCommandMeshEdit::QgsMeshLayerUndoCommandMeshEdit( QgsMeshEditor *meshEditor ):
+  mMeshEditor( meshEditor )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:263]:
```suggestion
QgsMeshLayerUndoCommandMeshEdit::QgsMeshLayerUndoCommandMeshEdit( QgsMeshEditor *meshEditor )
: mMeshEditor( meshEditor )
```

File: src/core/mesh/qgsmesheditor.cpp [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+{
+  if ( mMeshEditor.isNull() )
+    return;
+
+  for ( QgsMeshEditor::Edit &edit : mEdits )
+    mMeshEditor->applyEdit( edit );
+}
+
+QgsMeshLayerUndoCommandAddVertices::QgsMeshLayerUndoCommandAddVertices( QgsMeshEditor *meshEditor, const QVector<QgsMeshVertex> &vertices ):
+  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:286]:
```suggestion
QgsMeshLayerUndoCommandAddVertices::QgsMeshLayerUndoCommandAddVertices( QgsMeshEditor *meshEditor, const QVector<QgsMeshVertex> &vertices )
:  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```

File: src/core/mesh/qgsmesheditor.cpp [hunk 20]
```diff
@@ -0,0 +1,383 @@
...
+  }
+  else
+  {
+    for ( QgsMeshEditor::Edit &edit : mEdits )
+      mMeshEditor->applyEdit( edit );
+  }
+}
+
+QgsMeshLayerUndoCommandRemoveVertices::QgsMeshLayerUndoCommandRemoveVertices( QgsMeshEditor *meshEditor, const QList<int> &verticesToRemoveIndexes, bool fillHole ):
+  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:313]:
```suggestion
QgsMeshLayerUndoCommandRemoveVertices::QgsMeshLayerUndoCommandRemoveVertices( QgsMeshEditor *meshEditor, const QList<int> &verticesToRemoveIndexes, bool fillHole )
 :  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```

File: src/core/mesh/qgsmesheditor.cpp [hunk 21]
```diff
@@ -0,0 +1,383 @@
...
+  }
+  else
+  {
+    for ( QgsMeshEditor::Edit &edit : mEdits )
+      mMeshEditor->applyEdit( edit );
+  }
+}
+
+QgsMeshLayerUndoCommandAddFaces::QgsMeshLayerUndoCommandAddFaces( QgsMeshEditor *meshEditor, QgsTopologicalMesh::TopologicalFaces &faces ):
+  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:338]:
```suggestion
QgsMeshLayerUndoCommandAddFaces::QgsMeshLayerUndoCommandAddFaces( QgsMeshEditor *meshEditor, QgsTopologicalMesh::TopologicalFaces &faces )
:  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```

File: src/core/mesh/qgsmesheditor.cpp [hunk 22]
```diff
@@ -0,0 +1,383 @@
...
+  }
+  else
+  {
+    for ( QgsMeshEditor::Edit &edit : mEdits )
+      mMeshEditor->applyEdit( edit );
+  }
+}
+
+QgsMeshLayerUndoCommandRemoveFaces::QgsMeshLayerUndoCommandRemoveFaces( QgsMeshEditor *meshEditor, const QList<int> &facesToRemoveIndexes ):
+  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.cpp:360]:
```suggestion
QgsMeshLayerUndoCommandRemoveFaces::QgsMeshLayerUndoCommandRemoveFaces( QgsMeshEditor *meshEditor, const QList<int> &facesToRemoveIndexes )
:  QgsMeshLayerUndoCommandMeshEdit( meshEditor )
```

File: src/core/mesh/qgsmesheditor.h [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+ * \brief Class that represents an error during mesh editing
+ *
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsMeshEditingError
+{
+  public:
+
+    //! Type of the error
+    enum MeshEditingErrorType
```
Comment (nyalldawson) [inline src/core/mesh/qgsmesheditor.h:40]: Can you please move this to the Qgis class, and promote it to an enum class (also add docs for the individual member values)? We're moving toward having all enums in a common class (Qt style) to easily facilitate forward declaration vs header imports.

File: src/core/mesh/qgstopologicalmesh.cpp [hunk 24]
```diff
@@ -0,0 +1,1132 @@
...
+static int vertexPositionInFace( const QgsMesh &mesh, int vertexIndex, int faceIndex )
+{
+  if ( faceIndex < 0 || faceIndex >= mesh.faceCount() )
+    return -1;
+
+  return vertexPositionInFace( vertexIndex, mesh.face( faceIndex ) );
+}
+
+QgsMeshVertexCirculator::QgsMeshVertexCirculator( const QgsTopologicalMesh &topologicalMesh, int vertexIndex ):
+  mFaces( topologicalMesh.mMesh->faces )
```
Comment (nyalldawson) [inline src/core/mesh/qgstopologicalmesh.cpp:37]:
```suggestion
QgsMeshVertexCirculator::QgsMeshVertexCirculator( const QgsTopologicalMesh &topologicalMesh, int vertexIndex )
:  mFaces( topologicalMesh.mMesh->faces )
```

File: src/core/mesh/qgstopologicalmesh.cpp [hunk 25]
```diff
@@ -0,0 +1,1132 @@
...
+  {
+    mIsValid = false;
+  }
+
+  if ( mIsValid )
+    mLastValidFace = mCurrentFace;
+}
+
+QgsMeshVertexCirculator::QgsMeshVertexCirculator( const QgsTopologicalMesh::TopologicalFaces &topologicalFaces, int faceIndex, int vertexIndex ):
+  mFaces( topologicalFaces.mFaces )
```
Comment (nyalldawson) [inline src/core/mesh/qgstopologicalmesh.cpp:56]:
```suggestion
QgsMeshVertexCirculator::QgsMeshVertexCirculator( const QgsTopologicalMesh::TopologicalFaces &topologicalFaces, int faceIndex, int vertexIndex )
:  mFaces( topologicalFaces.mFaces )
```

(elsewhere too!)

File: src/core/mesh/qgsmeshdataset.h (hunk 15 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshdataset.h:794]: intentional, not exposed at all

File: src/core/mesh/qgsmeshdataset.h (hunk 12 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshdataset.h:770]: intentional, not exposed at all

File: src/core/mesh/qgsmesheditor.h (hunk 23 above)
Comment (vcloarec) [reply src/core/mesh/qgsmesheditor.h]: done

**Review by PeterPetrik — APPROVED** [review]

File: src/core/mesh/qgstopologicalmesh.h (hunk 4 above)
Comment (PeterPetrik) [reply src/core/mesh/qgstopologicalmesh.h:226]: I prefer to have any more substantial class in separate file, but it is not requirement, just my preference/wish

File: src/core/mesh/qgsmeshdataset.h (hunk 12 above)
Comment (PeterPetrik) [reply src/core/mesh/qgsmeshdataset.h:770]: detto

**Pull request merged**